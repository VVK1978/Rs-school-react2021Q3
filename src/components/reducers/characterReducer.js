const SET_CHARACTERS = 'SET_CHARACTERS';
const SET_IS_FETCHING = 'SET_IS_FETCHING';
const SET_PER_PAGE = 'SET_PER_PAGE';
const SET_CURRENT_PAGE = 'SET_CURRENT_PAGE';
const SET_SEARCH_VALUE = 'SET_SEARCH_VALUE';
const SET_SORT_VALUE = 'SET_SORT_VALUE';
const SET_SORT_DIRECTION = 'SET_SORT_DIRECTION';

const defaultState = {
  items: [],
  isFetching: true,
  currentPage: 1,
  perPage: 10,
  totalPages: 0,
  searchValue: '',
  sortBy: 'name',
  sortDirection: 'asc',
};
export default function characterReducer(state = defaultState, action) {
  switch (action.type) {
    case SET_CHARACTERS:
      return {
        ...state,
        items: action.payload.docs,
        totalPages: action.payload.pages,
        isFetching: false,
      };
    case SET_IS_FETCHING:
      return {
        ...state,
        isFetching: action.payload,
      };
    case SET_CURRENT_PAGE:
      return {
        ...state,
        currentPage: action.payload,
      };
    case SET_PER_PAGE:
      return {
        ...state,
        perPage: action.payload,
      };
    case SET_SEARCH_VALUE:
      return {
        ...state,
        searchValue: action.payload,
      };
    case SET_SORT_VALUE:
      return {
        ...state,
        sortBy: action.payload,
      };
    case SET_SORT_DIRECTION:
      return {
        ...state,
        sortDirection: action.payload,
      };
    default:
      return state;
  }
}

export const setCharacters = (characters) => ({
  type: SET_CHARACTERS,
  payload: characters,
});

export const setIsFetching = (bool) => ({
  type: SET_IS_FETCHING,
  payload: bool,
});

export const setCurrentPage = (page) => ({
  type: SET_CURRENT_PAGE,
  payload: page,
});

export const setPerPage = (perPage) => ({
  type: SET_PER_PAGE,
  payload: perPage,
});

export const setSearchValue = (searchValue) => ({
  type: SET_SEARCH_VALUE,
  payload: searchValue,
});

export const setSort = (sortBy) => ({
  type: SET_SORT_VALUE,
  payload: sortBy,
});

export const setSortDirection = (sortDirection) => ({
  type: SET_SORT_DIRECTION,
  payload: sortDirection,
});
