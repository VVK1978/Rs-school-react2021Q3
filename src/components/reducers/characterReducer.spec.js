/* eslint-disable no-undef */
import configureStore from 'redux-mock-store';
import * as defaultState from './characterReducer';
import characterReducer from './characterReducer';

describe('Reducer tests on dispatching', () => {
  const mockStore = configureStore(defaultState);
  let store;

  beforeEach(() => {
    store = mockStore(defaultState);
  });

  it('should checking action SET_PER_PAGE on dispatching', () => {
    store.dispatch(defaultState.setPerPage(20));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_PER_PAGE');
  });

  it('should checking action SET_PER_PAGE', () => {
    const action = {
      type: 'SET_PER_PAGE',
      payload: 20,
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      perPage: action.payload,
    });
  });

  it('should checking action SET_IS_FETCHING on dispatching', () => {
    store.dispatch(defaultState.setIsFetching(true));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_IS_FETCHING');
  });

  it('should checking action SET_IS_FETCHING', () => {
    const action = {
      type: 'SET_IS_FETCHING',
      payload: true,
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      isFetching: action.payload,
    });
  });

  it('should checking action SET_CURRENT_PAGE on dispatching', () => {
    store.dispatch(defaultState.setCurrentPage(2));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_CURRENT_PAGE');
  });

  it('should checking action SET_CURRENT_PAGE', () => {
    const action = {
      type: 'SET_CURRENT_PAGE',
      payload: 2,
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      currentPage: action.payload,
    });
  });

  it('should checking action SET_SEARCH_VALUE on dispatching', () => {
    store.dispatch(defaultState.setSearchValue('2'));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_SEARCH_VALUE');
  });

  it('should checking action SET_SEARCH_VALUE', () => {
    const action = {
      type: 'SET_SEARCH_VALUE',
      payload: 2,
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      searchValue: action.payload,
    });
  });

  it('should checking action SET_SORT_VALUE on dispatching', () => {
    store.dispatch(defaultState.setSort('name'));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_SORT_VALUE');
  });

  it('should checking action SET_SORT_VALUE', () => {
    const action = {
      type: 'SET_SORT_VALUE',
      payload: 'name',
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      sortBy: action.payload,
    });
  });

  it('should checking action SET_SORT_DIRECTION on dispatching', () => {
    store.dispatch(defaultState.setSortDirection('asc'));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_SORT_DIRECTION');
  });

  it('should checking action SET_SORT_DIRECTION', () => {
    const action = {
      type: 'SET_SORT_DIRECTION',
      payload: 'asc',
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      sortDirection: action.payload,
    });
  });

  it('should checking action SET_CHARACTERS on dispatching', () => {
    store.dispatch(defaultState.setCharacters([{ a: 1, b: 2 }]));
    const action = store.getActions();
    expect(action[0].type).toBe('SET_CHARACTERS');
  });

  it('should checking action SET_CHARACTERS', () => {
    const action = {
      type: 'SET_CHARACTERS',
      payload: { docs: { a: 1, b: 2 }, pages: 5 },
      isFetching: false,
    };
    expect(characterReducer(defaultState, action)).toEqual({
      ...defaultState,
      items: action.payload.docs,
      totalPages: action.payload.pages,
      isFetching: action.isFetching,
    });
  });
});
