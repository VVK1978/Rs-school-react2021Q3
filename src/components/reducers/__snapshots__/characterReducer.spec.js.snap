// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer tests on dispatching should checking action SET_PER_PAGE snapshot 1`] = `
Object {
  "currentPage": 1,
  "isFetching": true,
  "items": Array [],
  "perPage": 10,
  "searchValue": "",
  "sortBy": "name",
  "sortDirection": "asc",
  "totalPages": 0,
}
`;
